### PR TITLE
[WIP]Experimental pluggable deltas

### PR DESCRIPTION
**Pull request by codemerx**
Provides the infrastructure in the client library that can handle pluggable codecs as specified in https://github.com/ably/wiki/issues/380#issuecomment-556189239. Also adds the capability to work with the `vcdiff` pluggable decoder defined in https://github.com/ably/ably-delta-codec-java/pull/3.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "lib/src/test/resources/ably-common"]
 	path = lib/src/test/resources/ably-common
 	url = https://github.com/ably/ably-common.git
+[submodule "submodules/ably-delta-codec-java"]
```
Comment (mattheworiordan) [inline .gitmodules]: OOI, what is the thinking about this being a submodule?

Comment (codemerx) [reply .gitmodules]: We need this for the delta tests. The library does not refer it directly and has no knowledge of it but the tests have to be able to create the pluggable `vcdiff` decoder and provide it to the library instance.

File: lib/src/main/java/io/ably/lib/types/BaseMessage.java [hunk 2]
```diff
@@ -110,13 +123,37 @@ public void decode(ChannelOptions opts) throws MessageDecodeException {
...
+								try {
+									data = vcdiffCodec.decode((byte[]) data, context.get_lastMessage());
+									lastPayload = data;
+								}
+								catch (AblyException ex) {
+									throw MessageDecodeException.fromThrowableAndErrorInfo(ex, new ErrorInfo("Decoding failed for user provided codec " + match.group(1), ex.errorInfo.statusCode, ex.errorInfo.code));
+								}
+
+								continue;
+							}
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/BaseMessage.java]: If the codec is not present control will fall out of the switch statement (`continue` won't get executed) and `break` will get executed and the decoding loop will be over. So teh message will be returned decoded up to but not including `vcdiff` and the decoding string will contain everything up to and including `vcdiff`. That's what the library does when it encounters encoding it  cannot handle.

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java [hunk 3]
```diff
@@ -194,4 +195,6 @@ public ClientOptions(String key) throws AblyException {
 	 * before responding.
 	 */
 	public boolean pushFullWait = false;
+
+	public final Map<String, PluggableCodec> Codecs = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: We do not have plugins any more. We gave up on introducing generic plugins. What we pass to the lib here is codecs indeed. They should be able to decode and maybe encode in particular format.

[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: The way it is now client code using "new" codecs would work against older versions of the libraries that do not support these, too. As mentioned above they will just decode what is possible to decode and return the intermediate result. Customer code can then decode a step and return the rest to the lib for further decoding. Having an enum we lose that workflow.

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	private byte[] lastMessageBinary;
+
+	public DecodingContext(Map<String, PluggableCodec> codecs)
+	{
+		lastMessageBinary = null;
+		lastMessageString = null;
+		this.codecs = codecs;
+	}
+
+	public Map<String, PluggableCodec> codecs;
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: I personally am not a fan of public fields in languages like Java. In this project, however, we have them all over the place so I am adhering to the existing coding style. Changing it piecemeal would probably cause bigger mess.

I am not really sure I understand why you think that field does not belong to the `DecodingContext` class. Can you elaborate? Where does it belong?

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+	public byte[] get_lastMessage() {
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: 1. That's a bug. Thanks for pointing it out. Fixed now.

2. We are setting this on each string message the library receives on each channel. We are getting this only if next message is delta. For performance reasons converting to utf-8 is done only when needed.

The reasoning behind this class design (including UTF-8)  is documented [in the PR description](https://github.com/ably/ably-java/pull/532#issue-345383943).

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+		if(lastMessageBinary != null)
+			return lastMessageBinary;
+		else if(lastMessageString != null) {
+			return lastMessageString.getBytes(StandardCharsets.UTF_8);
+		}
+		else
+			return null;
+	}
+
+	public void set_lastMessage(String message) {
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: True that. Fixed now.

File: lib/src/main/java/io/ably/lib/types/PluggableCodec.java [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package io.ably.lib.types;
+
+public interface PluggableCodec {
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/PluggableCodec.java]: Replied in previous comment.

File: lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package io.ably.lib.types;
+
+public interface VCDiffPluggableCodec extends PluggableCodec {
+	byte[] decode(byte[] delta, byte[] base) throws AblyException;
```
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java]: 1. These are all valid questions but they apply to pretty much any method in the library. Why would we document the invalid input behavior of this one in particular? 
2. I am adhering to the [spec](https://github.com/ably/wiki/issues/380#issuecomment-556189239) we previously agreed on. Not really sure the benefit we are going to get from reversing the arguments order at this point is worth the effort to produce another version of the spec, change code, run tests, etc.
3. Could you elaborate? What would be a better candidate?

File: lib/src/main/java/io/ably/lib/types/BaseMessage.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/BaseMessage.java]: Thanks for explaining.

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: Oh, ok... I know it's 'just' a name, however it's a key component of the public API. This differs to a verbal discussion I had with @mattheworiordan and @paddybyers a few days ago. As a point of extensibility, is there a reason why these things need to be restrained to performing the role of a codec?

File: lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java]: 1. Perhaps a discussion for another day :neutral_face:.
2. No change to the spec, just the presentation - I think the effort is worth it in this case - I can do that piece of work if that would help?
3. Now I look at it closer, I'm a bit torn. **On the one hand**, should it even be an `AblyException` at all? Looking at that hierarchy, they seem to be exceptions which get emitted by the library internally. **On the other hand**, however, we do already have a `MessageDecodeException` - so perhaps that would be more suitable for this particular method on this particular codec plugin?

Screenshot 2019-11-27 at 12 45 29

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: I'm not sure that's true, or at least not if we go with this being a 'plugin' rather than a 'codec'. Where, for a plugin, there's assumed to be more internal wiring required in a new release to make it work therefore it can't simply be supplied to an older library.

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: I get your point around existing, implicit coding style (i.e. inherited from surrounding code). However, I would rather that new code adheres to accepted Java conventions... in lieu of us producing an explicit coding style and/or adopting a tool like checkstyle.

The `codecs` field is just wide open. It's never actively used by this class and, on top of that, the reference to the map is mutable, as are the contents! Perhaps you could refine to what's really required? e.g. make the field private and then expose a public method `PluggableCodec findCodec(String codecName)` and define that if the codec is not found then `null` will be returned, which is then a decision for the calling code. What do you think?

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: ok... could it not be documented a little closer to home - in the code itself?
I guess it could be in the spec (which I know you're working on) but I'm guessing that optimisations like that are going to be specific to particular client library SDK implementation needs.

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: Well this extensibility point was planned as being able to accept codecs only as per https://github.com/ably/wiki/issues/380#issuecomment-556189239. In theory we could have extensibility point that accepts anything but this was never planned to be part of this effort. 

Shouldn't this really have been part of the discussion here - https://github.com/ably/wiki/issues/380#issuecomment-556189239 anyways? It looks like you approved this spec actually and that's where the `Codecs` name comes from.

File: lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java]: 1. We should have this discussion I guess. There seems to be little benefit in postponing it.
2. The infrastructure we have here is quite complicated atm. There are several levels of branches out of branches, no tests for corner cases, we have stuff like two repos each of which contains the other one as submodule and itself as a submodule of the submodule. Documentation is also  in various places and in various degrees of sync. You doing the work will take some knowledge transfer for sure and thus will take even more time combined. Not convinced the order of arguments in a two argument method is worth the effort. @paddybyers wdyt?
3. Yep, name is more descriptive (sadly contents are the same)

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: Currently it is a codec. If we want to make it a plugin then a lot of things will get very different. As mentioned above not really sure generic plugins should be introduced as part of this effort to replace the generic codecs that were also part of this effort and we gave up on.

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: As mentioned I personally prefer the architecture you describe. I still think that having drastically different standards for the new code would result in a bad mixture of coding styles but that decision is yours to make.

Before doing that change, though, let's see whether we are not missing something.  We have a lot of public collections all over the place (e.g. only public `Map`s are a couple of dozens) and maybe there is a reason for that. @paddybyers what would that be?

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: Well that should again be part of our discussion on how we choose which methods to document deeply. There is non-trivial, last mesage bookkeping like this in a couple of more places in the code. Why do we document this one but not the other ones?

[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: I just think some documentation is better than none, and documentation is always best written at the point the code it's documenting is written. As an average developer, I had to ask questions around these interfaces which suggests that others would find themselves in the same situation later. Given we have quite a number of people who maintain the codebase now and there will be more in time, we just need to make sure they get as much help as possible when they find themselves back in this place.

File: lib/src/main/java/io/ably/lib/types/DecodingContext.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/DecodingContext.java]: Yeah, @paddybyers and I have had quite a few discussions already around my observations of the Java codebase and where things are, perhaps, a little 'over exposed'. Those interfaces will have to stand as they are until we get the bandwidth to refactor / deprecate / etc.. In the meantime, however, we must strive to make any new interfaces clean and minimal - especially when they form part of our front line public API. That approach has got to prevent some bugs from making their way in to both our code and that of developers using our code.

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: Can you expand on what needs to be very different? The codec / plugin interface is empty so I thought this was just a matter of naming.

File: lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java]: I'm not too worried about corner case testing, branching or repository structure in respect of submodules at the moment as all these things are side effects of the development process.

My sole concern is what APIs customers will be using day to day once this API is released.

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: It should have so my bad. I just think, now that I've noticed it, it's worth addressing.

:innocent:

[the remark this answers is not in the document]
Comment (paddybyers) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: My recollection is that we agreed `plugins`.

File: .gitmodules (hunk 1 above)
Comment (mattheworiordan) [reply .gitmodules]: Ok. @paddybyers @QuintinWillison I thought you told me the other day the plan was now to bundle in VCDiff as part of the Java distribution.  Have I got my wires crossed?

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattheworiordan) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: I thought we agreed `plugins` too given we don't support `codecs`, we only support one `codec` anyway, so `plugins` gives us more extensibility in future.

File: .gitmodules (hunk 1 above)
Comment (QuintinWillison) [reply .gitmodules]: Good point, though I didn't think we had decided absolutely as we've not yet assessed what the impact is on library size for Android of baking VCDiff in. @codemerx, do you know how much it adds to the size?

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: This change is out of date as this point has been addressed. I'll resolve this thread.

File: lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply lib/src/main/java/io/ably/lib/types/VCDiffPluggableCodec.java]: I see this is now `MessageDecodeException`. Thanks.

File: .gitmodules (hunk 1 above)
Comment (codemerx) [reply .gitmodules]: The jar is about 108K.

File: lib/src/main/java/io/ably/lib/types/ClientOptions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (codemerx) [reply lib/src/main/java/io/ably/lib/types/ClientOptions.java]: I changed this to `plugins` yesterday. Are you guys looking at an older version of the code?

File: .gitmodules (hunk 1 above)
Comment (QuintinWillison) [reply .gitmodules]: So, with `ably-android-1.1.6.aar` at just over 340 KB, I think I am right in calculating the impact of 108 KB as an increase of about **32%**.

**Comment by paddybyers** [comment]
Closed in favour of https://github.com/ably/ably-java/pull/544
